### PR TITLE
refactor: Convert SchedV2 to Kotlin

### PR DESCRIPTION
**Pull request by david-allison**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 1]
```diff
@@ -17,265 +17,221 @@
...
+import com.ichi2.libanki.*
+import com.ichi2.libanki.Collection
+import com.ichi2.libanki.Consts.BUTTON_TYPE
+import com.ichi2.libanki.Consts.CARD_QUEUE
+import com.ichi2.libanki.Consts.DYN_PRIORITY
+import com.ichi2.libanki.Consts.NEW_CARD_ORDER
+import com.ichi2.libanki.Consts.REVLOG_TYPE
+import com.ichi2.libanki.SortOrder.AfterSqlOrderBy
+import com.ichi2.libanki.Utils.collection2Array
+import com.ichi2.libanki.Utils.ids2str
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:42]: Please revert the imports for these two methods from Utils.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 2]
```diff
@@ -285,581 +241,564 @@ public class SchedV2 extends AbstractSched {
...
+        if (!BackendFactory.defaultLegacySchema) {
+            return super.quickDeckDueTree()
         }
 
         // Similar to deckDueList
-        ArrayList<DeckTreeNode> allDecksSorted = new ArrayList<>();
-        for (JSONObject deck : getCol().getDecks().allSorted()) {
-            DeckTreeNode g = new DeckTreeNode(deck.getString("name"), deck.getLong("id"));
-            allDecksSorted.add(g);
+        val allDecksSorted = ArrayList<DeckTreeNode>()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:520]: @KotlinCleanup("simplify with map {}")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 3]
```diff
@@ -1152,339 +1089,335 @@ public class SchedV2 extends AbstractSched {
...
         // if we were dynamic, graduating means moving back to the old deck
-        if (card.isInDynamicDeck()) {
-            _removeFromFiltered(card);
+        if (card.isInDynamicDeck) {
+            _removeFromFiltered(card)
         }
     }
 
-    private void _rescheduleGraduatingLapse(@NonNull Card card, boolean early) {
+    private fun _rescheduleGraduatingLapse(card: Card, early: Boolean) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:1282]: @KotlinCleanup("simplify with scope function on card")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 4]
```diff
@@ -1152,339 +1089,335 @@ public class SchedV2 extends AbstractSched {
...
-    protected void _logLrn(@NonNull Card card, @Consts.BUTTON_TYPE int ease, @NonNull JSONObject conf, boolean leaving, @Consts.REVLOG_TYPE int type, int lastLeft) {
-        int lastIvl = -(_delayForGrade(conf, lastLeft));
-        int ivl = leaving ? card.getIvl() : -(_delayForGrade(conf, card.getLeft()));
-        log(card.getId(), getCol().usn(), ease, ivl, lastIvl, card.getFactor(), card.timeTaken(), type);
-    }
-
-    protected void log(long id, int usn, @Consts.BUTTON_TYPE int ease, int ivl, int lastIvl, int factor, int timeTaken, @Consts.REVLOG_TYPE int type) {
+     * Overridden: V1 does not set type and queue */
+    private fun _rescheduleNew(card: Card, conf: JSONObject, early: Boolean) {
+        card.ivl = _graduatingIvl(card, conf, early)
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:1353]: @KotlinCleanup("simplify with scope function on card")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 5]
```diff
@@ -1523,398 +1455,354 @@ public class SchedV2 extends AbstractSched {
...
-    protected void _resetRevCount(@Nullable CancelListener cancelListener) {
-        int lim = _currentRevLimit(true);
-        if (isCancelled(cancelListener)) return;
-        mRevCount = getCol().getDb().queryScalar("SELECT count() FROM (SELECT id FROM cards WHERE did in " + _deckLimit() + " AND queue = " + Consts.QUEUE_TYPE_REV + " AND due <= ? AND id != ? LIMIT ?)",
-                                             mToday, currentCardId(), lim);
+    protected open fun _resetRevCount() {
+        _resetRevCount(null)
     }
 
+    protected open fun _resetRevCount(cancelListener: CancelListener?) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:1502]: @KotlinCleanup("see if the two versions of this function can be combined")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 6]
```diff
@@ -1924,884 +1812,813 @@ public class SchedV2 extends AbstractSched {
...
+    protected open fun _newConf(card: Card): JSONObject {
+        val conf = _cardConf(card)
+        if (!card.isInDynamicDeck) {
+            return conf.getJSONObject("new")
         }
         // dynamic deck; override some attributes, use original deck for others
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONObject dict = new JSONObject();
+        val oconf = col.decks.confForDid(card.oDid)
+        val dict = JSONObject()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:1928]: @KotlinCleanup("simplify with return and scope function on dict")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt [hunk 7]
```diff
@@ -1924,884 +1812,813 @@ public class SchedV2 extends AbstractSched {
...
+    protected open fun _lapseConf(card: Card): JSONObject {
+        val conf = _cardConf(card)
+        if (!card.isInDynamicDeck) {
+            return conf.getJSONObject("lapse")
         }
         // dynamic deck; override some attributes, use original deck for others
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONObject dict = new JSONObject();
+        val oconf = col.decks.confForDid(card.oDid)
+        val dict = JSONObject()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.kt:1949]: @KotlinCleanup("simplify with return and scope function on dict")

**Review by lukstbit — CHANGES_REQUESTED** [review]
Finally managed to review this, looks good with some minor cleanups.

**Review by lukstbit — APPROVED** [review]
LGTM

**Review by mikehardy — APPROVED** [review]
incredible

**Pull request merged**